src/internal/Observable.ts

```diff
@@ -318,7 +318,7 @@ export class Observable<T> implements Subscribable<T> {
    *   .subscribe(x => console.log(x))
    * ```
    */
-  pipe<R>(...operations: OperatorFunction<T, R>[]): Observable<R> {
+  pipe<R>(...operations: OperatorFunction<any, any>[]): Observable<R> {
```

Comment:
One thing to consider here is `pipe<R=any>`, so that you don't over-rely on return type for inference.

Reply:
@rkirov Can the inference really bubble out of the implementation? I'll take your word for it, as I'm unclear on the actual inference rules. (If you know of any resources - other than the TypeScript source - that explain the workings of TypeScript's type inference, I'd really appreciate if you could direct me to them.)

Anyway, I'm not sure that `R` is even required here, shouldn't we use something like this?

```ts
pipe(...operations: OperatorFunction<any, any>[]): Observable<any>
```

With the `any` type parameters already being specified in the rest-parameters type and with it being the implementation of an overloaded function, does `R` really offer any value whatsoever?

Reply:
You are right about "bubbling out", my comment was pertaining to line 299, but I got mixed up. The types here are only relevant for 'implementation', not for the consumer.

About learning how type inference works - TS doesn't maintain a spec any longer, so the best one can do if they want to learn the rules is keep testing with small examples (what I do) or read the source code (haven't done that for type inference).

What makes matters hard here is that this is mixing generics and function overloads. Separately each one has odd corners, and together things are twice as hard.

Historically, we (team maintaining TS in google) have been having a lot of issues with generics that are only inferred through the return type of a function - simplest example being `f<T>(): T`. TS will try to use an context type to pick T, but sometimes there will be no such type to pick, so it will fallback to `{}`. 

What happens is that in simple cases everything works fine:
```typescript
const x: string = f();  // T is inferred to be string
const y = f();  // y is {}, which is ok, its an opaque object.
```

But what we see the inferred `{}` can be propagating too much, and into another loosely typed API.
```typescript
function g() {
  return f();
}
function someAPI(x: any) {...}

someAPI(g());
```

Now, if the maintainer tries to change `someAPI` to be stricter say 'x: string', they are confronted with a `{}` that flowed through without anybody noticing, which makes it hard to change.  That's why we are considering whether it would have been better if the fallback for when TS cannot infer the generic type should have been `any` instead of the default `{}` (technically it is the lower bound, i.e whatever is after `extends`). Paradoxically, even though `any` is less safe, it can help making other type change stricted.

Maybe it would have been better if TS forced people to write a type like `f<Something>`, instead of fallback on `{}`.

We are still figuring out what is most ergonomic for ppl, would love to hear your opinion too.

Reply:
> Maybe it would have been better if TS forced people to write a type like `f<Something>`, instead of fallback on `{}`.

Yes. I agree wholeheartedly with this.